### PR TITLE
separate kube and openshift admission

### PR DESCRIPTION
**Pull request by deads2k**
This separates kube and openshift admission and starts basing kube admission on upstream kube.

/assign @soltysh @sttts

### CODE REVIEW

File: pkg/configconversion/simple.go [hunk 1]
```diff
@@ -7,7 +7,9 @@ import (
 	openshiftcontrolplanev1 "github.com/openshift/api/openshiftcontrolplane/v1"
 	osinv1 "github.com/openshift/api/osin/v1"
 	"github.com/openshift/library-go/pkg/crypto"
-	"github.com/openshift/origin/pkg/apiserver/admission"
+	"github.com/openshift/origin/pkg/cmd/openshift-apiserver/openshiftadmission"
+	"github.com/openshift/origin/pkg/cmd/openshift-kube-apiserver/kubeadmission"
+	"k8s.io/kubernetes/pkg/kubeapiserver/options"
```
Comment (soltysh) [inline pkg/configconversion/simple.go]: Move up.

File: pkg/cmd/openshift-kube-apiserver/server.go [hunk 2]
```diff
@@ -5,17 +5,16 @@ import (
 
 	"github.com/golang/glog"
 
-	"k8s.io/apimachinery/pkg/util/sets"
+	kubecontrolplanev1 "github.com/openshift/api/kubecontrolplane/v1"
```
Comment (soltysh) [inline pkg/cmd/openshift-kube-apiserver/server.go]: Down.

File: pkg/cmd/openshift-kube-apiserver/kubeadmission/register.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+		"ProjectRequestLimit",
+		"RunOnceDuration",
+		"PodNodeConstraints",
+		overrideapi.PluginName,
+		imagepolicyapi.PluginName,
+		"openshift.io/RestrictSubjectBindings",
+	)
+)
+
+func NewOrderedKubeAdmissionPlugins(kubeAdmissionOrder []string) []string {
```
Comment (soltysh) [inline pkg/cmd/openshift-kube-apiserver/kubeadmission/register.go]: Shouldn't we have a test verifying that our methods are in-sync with upstream list of admissions? This applies for both for `NewOrderedKubeAdmissionPlugins` and `NewDefaultOffPluginsFunc`

File: pkg/cmd/openshift-apiserver/openshiftadmission/register.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	schedulerpodnodeconstraints.Register(plugins)
+	imagepolicy.Register(plugins)
+	quotaclusterresourcequota.Register(plugins)
+}
+
+var (
+	// OpenShiftAdmissionPlugins gives the in-order default admission chain for openshift resources.
+	OpenShiftAdmissionPlugins = []string{
+		lifecycle.PluginName,
+		"OwnerReferencesPermissionEnforcement",
```
Comment (soltysh) [inline pkg/cmd/openshift-apiserver/openshiftadmission/register.go]: I wonder if we should have additional test for OpenShift admissions to ensure the hardcoded names are in-sync with upsream?

File: pkg/cmd/openshift-apiserver/openshiftapiserver/config.go [hunk 5]
```diff
@@ -156,19 +156,11 @@ func NewOpenshiftAPIConfig(config *openshiftcontrolplanev1.OpenShiftAPIServerCon
 	clusterQuotaMappingController := NewClusterQuotaMappingController(informers.kubernetesInformers.Core().V1().Namespaces(), informers.quotaInformers.Quota().InternalVersion().ClusterResourceQuotas())
 	discoveryClient := cacheddiscovery.NewMemCacheClient(kubeClient.Discovery())
 	restMapper := restmapper.NewDeferredDiscoveryRESTMapper(discoveryClient)
-	admissionInitializer, err := originadmission.NewPluginInitializer(config.ImagePolicyConfig.ExternalRegistryHostnames, config.ImagePolicyConfig.InternalRegistryHostname, config.CloudProviderFile, kubeClientConfig, informers, genericConfig.Authorization.Authorizer, projectCache, restMapper, clusterQuotaMappingController)
+	admissionInitializer, err := openshiftadmission.NewPluginInitializer(config.ImagePolicyConfig.ExternalRegistryHostnames, config.ImagePolicyConfig.InternalRegistryHostname, config.CloudProviderFile, kubeClientConfig, informers, genericConfig.Authorization.Authorizer, projectCache, restMapper, clusterQuotaMappingController)
 	if err != nil {
 		return nil, err
 	}
-	namespaceLabelDecorator := namespaceconditions.NamespaceLabelConditions{
```
[the remark this answers is not in the document]
Comment (deads2k) [reply pkg/cmd/openshift-apiserver/openshiftapiserver/config.go]: > I'm guessing we're falling on to upstream one?

the openshift apiserver never tries to run admission before admission is ready, so we don't have to skip ourselves.

File: pkg/cmd/openshift-kube-apiserver/kubeadmission/register.go (hunk 3 above)
Comment (deads2k) [reply pkg/cmd/openshift-kube-apiserver/kubeadmission/register.go]: > Shouldn't we have a test verifying that our methods are in-sync with upstream list of admissions? This applies for both for `NewOrderedKubeAdmissionPlugins` and `NewDefaultOffPluginsFunc`

I don't think so.  I the idea of this pull is that we always use the upstream order and only inject in a couple spots.  I think I'm even going to change those spots to have a single injection point in a followup pull.  If you couldn't be a webhook, you can't be included.

File: pkg/cmd/openshift-apiserver/openshiftadmission/register.go (hunk 4 above)
Comment (deads2k) [reply pkg/cmd/openshift-apiserver/openshiftadmission/register.go]: > I wonder if we should have additional test for OpenShift admissions to ensure the hardcoded names are in-sync with upsream?

Pretty sure we will hard fail if we put an invalid item in the list

File: pkg/cmd/openshift-apiserver/openshiftadmission/chain_builder.go [hunk 6]
```diff
@@ -1,4 +1,4 @@
-package admission
+package openshiftadmission
```
[the remark this answers is not in the document]
Comment (soltysh) [reply pkg/cmd/openshift-apiserver/openshiftadmission/chain_builder.go]: Yet, we always alias them... I still don't feel convinced but I won't block on that.

**Comment by deads2k** [comment]
serial failure is known flake
`openshift-tests [sig-api-machinery] Namespaces [Serial] should ensure that all pods are removed when a namespace is deleted [Conformance] [Suite:openshift/conformance/serial/minimal] [Suite:k8s`

**Comment by deads2k** [comment]
@gabemontero can you have a look at the builds failure?  I'm seeing
`Feb  4 16:51:03.098: INFO: At 2019-02-04 16:48:48 +0000 UTC - event for mys2itest-1-rxhrb: {kubelet ip-10-0-149-208.ec2.internal} Failed: Error: ErrImagePull
Feb  4 16:51:03.098: INFO: At 2019-02-04 16:48:48 +0000 UTC - event for mys2itest-1-rxhrb: {kubelet ip-10-0-149-208.ec2.internal} Failed: Failed to pull image "image-registry.openshift-image-registry.svc:5000/e2e-test-cli-test-hooks-dqlxk/mys2itest@sha256:efa91a44ab95e4ad726a17adc2a6a2ed30148b24382ebf809aaa4e1723ec5ee4": rpc error: code = Unknown desc = Manifest does not match provided manifest digest `
which looks weird and I don't see how this change could have done that.

**Comment by deads2k** [comment]
aws-all looks to be 90% failure https://openshift-gce-devel.appspot.com/builds/origin-ci-test/pr-logs/directory/pull-ci-openshift-origin-master-e2e-aws-all

**Comment by deads2k** [comment]
conformance hasn't worked since christmas: https://openshift-gce-devel.appspot.com/builds/origin-ci-test/pr-logs/directory/pull-ci-openshift-origin-master-e2e-conformance-k8s

**Comment by deads2k** [comment]
So that leaves test-cmd.  On a green test-cmd I plan to merge.

**Comment by deads2k** [comment]
test only required failing job is a known perf flake. merging to unstick the rest of admission.

**Pull request merged**